processing/src/main/java/io/druid/query/groupby/epinephelinae/EpiGroupByStrategy.java

```diff
@@ -0,0 +1,205 @@
+import com.metamx.common.guava.Sequences;
+import com.metamx.common.guava.nary.BinaryFn;
+import io.druid.collections.BlockingPool;
+import io.druid.collections.StupidPool;
+import io.druid.data.input.MapBasedRow;
+import io.druid.data.input.Row;
+import io.druid.granularity.AllGranularity;
+import io.druid.granularity.QueryGranularity;
+import io.druid.guice.annotations.Global;
+import io.druid.guice.annotations.Merging;
+import io.druid.guice.annotations.Smile;
+import io.druid.query.DruidProcessingConfig;
+import io.druid.query.Query;
+import io.druid.query.QueryRunner;
+import io.druid.query.QueryWatcher;
+import io.druid.query.ResultMergeQueryRunner;
+import io.druid.query.aggregation.PostAggregator;
+import io.druid.query.groupby.GroupByQuery;
+import io.druid.query.groupby.GroupByQueryConfig;
+import io.druid.query.groupby.strategy.GroupByStrategy;
+import io.druid.query.groupby.strategy.GroupByStrategySelector;
+import io.druid.segment.StorageAdapter;
+import org.joda.time.DateTime;
+
+import java.nio.ByteBuffer;
+import java.util.Map;
+
+public class EpiGroupByStrategy implements GroupByStrategy
+{
+  static final String CTX_KEY_FUDGE_TIMESTAMP = "fudgeTimestamp";
+
+  private final DruidProcessingConfig processingConfig;
+  private final Supplier<GroupByQueryConfig> configSupplier;
+  private final StupidPool<ByteBuffer> bufferPool;
+  private final BlockingPool<ByteBuffer> mergeBufferPool;
+  private final ObjectMapper spillMapper;
+  private final QueryWatcher queryWatcher;
+
+  @Inject
+  public EpiGroupByStrategy(
+      DruidProcessingConfig processingConfig,
+      Supplier<GroupByQueryConfig> configSupplier,
+      @Global StupidPool<ByteBuffer> bufferPool,
+      @Merging BlockingPool<ByteBuffer> mergeBufferPool,
+      @Smile ObjectMapper spillMapper,
+      QueryWatcher queryWatcher
+  )
+  {
+    this.processingConfig = processingConfig;
+    this.configSupplier = configSupplier;
+    this.bufferPool = bufferPool;
+    this.mergeBufferPool = mergeBufferPool;
+    this.spillMapper = spillMapper;
+    this.queryWatcher = queryWatcher;
+  }
+
+  @Override
+  public Sequence<Row> mergeResults(
+      final QueryRunner<Row> baseRunner,
+      final GroupByQuery query,
+      final Map<String, Object> responseContext
+  )
+  {
+    // Merge streams using ResultMergeQueryRunner, then apply postaggregators, then apply limit (which may
+    // involve materialization)
+
+    final ResultMergeQueryRunner<Row> mergingQueryRunner = new ResultMergeQueryRunner<Row>(baseRunner)
+    {
+      @Override
+      protected Ordering<Row> makeOrdering(Query<Row> queryParam)
+      {
+        return ((GroupByQuery) queryParam).getRowOrdering(true);
+      }
+
+      @Override
+      protected BinaryFn<Row, Row, Row> createMergeFn(Query<Row> queryParam)
+      {
+        return new EpiGroupByBinaryFn((GroupByQuery) queryParam);
+      }
+    };
+
+    // Fudge timestamp, maybe. Necessary to keep timestamps in sync across partial queries.
```

Comment:
really it's just that the IncrementalIndex did some weird stuff with timestamps, and the expected results in the groupBy tests depend on it.